proposals/0634-metro-esm-support.md

```diff
@@ -0,0 +1,95 @@
+---
+title: Improved support for ECMAScript modules
+author:
+- Moti Zilberman
+date: 2023-04-17
+---
+
+# RFC0634: Improved support for ECMAScript modules
+
+## Summary
+
+Enabling built-in support for ECMAScript modules in Metro for all React Native projects, fixing limitations in the current setup.
+
+## Motivation
+
+Today, [`metro-react-native-babel-preset`](https://github.com/facebook/metro/tree/main/packages/metro-react-native-babel-preset) includes the standard [Babel ESM-->CommonJS plugin](https://reactnative.dev/docs/ram-bundles-inline-requires#inline-requires) which compiles `import`, `export` etc into `require()`.
+
+There are a few problems in this model:
+
+* **Problem 1:** [Inline requires](https://reactnative.dev/docs/ram-bundles-inline-requires#inline-requires) do not work reliably for `import` declarations ([metro#909](https://github.com/facebook/metro/issues/909)). There is a separate, undocumented preset option, `lazyImportExportTransform`, that controls `import` inlining via the Babel plugin's [`lazy`](https://babeljs.io/docs/babel-plugin-transform-modules-commonjs#lazy) option.
+* **Problem 2:** `Platform` inlining doesn't work for the common case of `import {Platform} from 'react-native'` ([metro#800](https://github.com/facebook/metro/pull/800#issuecomment-1092434015)).
+* **Problem 3:** [Lazy bundling](https://github.com/react-native-community/discussions-and-proposals/blob/main/proposals/0605-lazy-bundling.md) does not work, since Babel transforms away `import()` calls before Metro can process them.
```

Comment:
I mentioned this in person but just to record it here - for `import()` specifically, could we theoretically preserve those (which are obviously "inline" already) to unblock lazy bundling without changing our approach top level `import`?